File: src/edictum_server/schemas/approvals.py
"""Schemas for the approval-queue endpoints."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

ApprovalStatusType = Literal["pending", "approved", "denied", "timeout"]


class CreateApprovalRequest(BaseModel):
    agent_id: str = Field(..., max_length=255)
    tool_name: str = Field(..., max_length=255)
    tool_args: dict[str, Any] | None = None
    message: str = Field(..., max_length=10_000)
    timeout: int = Field(default=300, ge=1, le=86400)
    timeout_effect: Literal["deny", "allow"] = "deny"
    decision_source: str | None = Field(default=None, max_length=255)
    contract_name: str | None = Field(default=None, max_length=255)


class ApprovalResponse(BaseModel):
    id: str
    status: str
    agent_id: str
    tool_name: str
    tool_args: dict[str, Any] | None = None
    message: str
    env: str
    timeout_seconds: int
    timeout_effect: str
    decision_source: str | None = None
    contract_name: str | None = None
    decided_by: str | None = None
    decided_at: datetime | None = None
    decision_reason: str | None = None
    decided_via: str | None = None
    created_at: datetime


class SubmitDecisionRequest(BaseModel):
    approved: bool
    reason: str | None = Field(default=None, max_length=2000)
    decided_via: str | None = Field(default=None, max_length=64)

File: dashboard/src/lib/api/approvals.ts
import { request } from "./client"

export interface ApprovalResponse {
  id: string
  status: "pending" | "approved" | "denied" | "timeout"
  agent_id: string
  tool_name: string
  tool_args: Record<string, unknown> | null
  message: string
  env: string
  timeout_seconds: number
  timeout_effect: "deny" | "allow"
  decided_by: string | null
  decided_at: string | null
  decision_reason: string | null
  decision_source: string | null
  contract_name: string | null
  decided_via: string | null
  created_at: string
}

export interface ApprovalFilters {
  status?: string
  agent_id?: string
  tool_name?: string
  env?: string
  limit?: number
  offset?: number
}

export function listApprovals(filters: ApprovalFilters = {}) {
  const params = new URLSearchParams()
  for (const [key, value] of Object.entries(filters)) {
    if (value !== undefined) params.set(key, String(value))
  }
  const qs = params.toString()
  return request<ApprovalResponse[]>(`/approvals${qs ? `?${qs}` : ""}`)
}

export function getApproval(id: string) {
  return request<ApprovalResponse>(`/approvals/${id}`)
}

export function submitDecision(
  id: string,
  approved: boolean,
  decidedBy?: string,
  reason?: string,
) {
  return request<ApprovalResponse>(`/approvals/${id}`, {
    method: "PUT",
    body: JSON.stringify({
      approved,
      decided_by: decidedBy ?? null,
      reason: reason ?? null,
    }),
  })
}
